external/avro/src/main/scala/org/apache/spark/sql/avro/SparkAvroDatumWriter.scala

```diff
@@ -0,0 +1,419 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.avro
+
+import java.nio.ByteBuffer
+
+import scala.collection.JavaConverters._
+
+import org.apache.avro.{LogicalTypes, Schema, SchemaBuilder}
+import org.apache.avro.Conversions.DecimalConversion
+import org.apache.avro.LogicalTypes.{TimestampMicros, TimestampMillis}
+import org.apache.avro.Schema.Type._
+import org.apache.avro.generic.GenericData.{EnumSymbol, Fixed}
+import org.apache.avro.generic.GenericDatumWriter
+import org.apache.avro.io.Encoder
+import org.apache.avro.util.Utf8
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{SpecializedGetters, SpecificInternalRow}
+import org.apache.spark.sql.catalyst.util.DateTimeUtils
+import org.apache.spark.sql.execution.datasources.DataSourceUtils
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.internal.SQLConf.LegacyBehaviorPolicy
+import org.apache.spark.sql.types._
+
+/**
+ * Writer to write data in catalyst format to avro encoders
+ *
+ * @param rootAvroType
+ * @param rootCatalystType
+ * @param nullable
+ * @tparam D
+ */
+class SparkAvroDatumWriter[D](rootAvroType: Schema, rootCatalystType: DataType, nullable: Boolean)
```

Comment:
This commit moves to using native writer for avro files. Previously we had AvroSerializer => GenericAvroDatumWriter => Encoder. Now SparkAvroDatumWriter writes directly to the Encoder. Eg, previously we would first fill up a HashMap in AvroSerializer and then GenericAvroDatumWriter (ReflectAvroDatumWriter before the first commit in this PR) will write it out to the Encoder. With this change, we write it out directly to the Encoder. Similarly resolving unions does not need to happen twice anymore.

Write time improvements can be found in the attached pdf on pg 3 under column D.